buildtest/buildsystem/base.py

```diff
@@ -0,0 +1,558 @@
+        self.result = {}
+        self.build_id = None
+        self.metadata = {}
+        self.config_file = config_file
+
+        # A builder is required to define the type attribute
+        if not hasattr(self, "type"):
+            sys.exit(
+                "A builder base is required to define the 'type' as a class variable"
+            )
+
+        # The recipe must be loaded as a dictionary
+        if not isinstance(recipe_config, dict):
+            sys.exit(
+                "Please load a recipe configuration before providing to the builder."
+            )
+
+        # The type must match the type of the builder
+        self.recipe = recipe_config
+        if self.recipe.get("type") != self.type:
+            sys.exit(
+                "Mismatch in type. Builder expects %s but found %s."
+                % (self.type, self.recipe.get("type"))
+            )
+
+    def __str__(self):
+        return "[builder-%s-%s]" % (self.type, self.name)
+
+    def __repr__(self):
+        return self.__str__()
+
+    def _create_test_folders(self):
+        """Create all needed test folders on init, and add their paths
+           to self.metadata
+        """
+        testpath = os.path.expandvars(self.metadata["testpath"])
+        testdir = os.path.dirname(testpath)
+        create_dir(testdir)
+        for folder in ["run", "log"]:
+            name = "%sdir" % folder
+            self.metadata[name] = os.path.join(testdir, folder)
+            create_dir(self.metadata[name])
+
+    def _init_logger(self, to_file=True):
+        """initialize the logger. This is called at the end of prepare_run,
+           so the testpath is defined with the build id.
+        """
+        if to_file:
+            self.metadata["logfile"] = os.path.join(
+                self.metadata["logdir"], "%s.log" % self.build_id
+            )
+            self.logger = init_logfile(self.metadata["logfile"])
+        else:
+            self.logger = init_log()
+
+    def get_test_extension(self):
+        """Return the test extension, which depends on the shell used. We
+           return .sh for all shell types except for python, we return .py.
+        """
+        shell = self.get_shell()
+        if "python" in shell:
+            return "py"
+        return "sh"
+
+    def get_environment(self):
+        """Take the environment section, return a list of lines to add
+           to the start of the testscript. Return lines that define
+           variables specific to the shell.
+
+           Returns: list of lines to add to beginning of test script.
+        """
+        env = []
+        pairs = self.recipe.get("env")
+        shell = self.get_shell()
+
+        # If we are using a python shell, we need to import os first
+        if "python" in shell:
+            env.append("import os")
+
+        # Parse environment depending on expected shell
+        if pairs:
+
+            # Handles csh and tsch
+            if "csh" in shell:
```

Comment:
might as well add **tcsh** in the if condition

Reply:
See the comment.